### PR TITLE
better fragment mode

### PR DESCRIPTION
**Pull request by yunfeiguo**
* in fragment mode, when `--estimate_np_by_length` enabled, sum up all sampled insert lengths and consider it as the length of polymerase read.
* break up the polymerase read length by assuming that first bp is in insert (not in adapter), i.e. the the number of passes is calculated based on length of insert (template) and adapter except for the first (maybe full or partial) pass.
* one test added

### CODE REVIEW

File: simulate.py [hunk 1]
```diff
@@ -61,6 +61,8 @@ def get_contigs(self):
     parser.add_argument("--max_pass", help="Maximum passes", type=int, default=1000000000)
     parser.add_argument("--output_polymerase_read", help="shall we simulate polymerase read (including adapter)?", action='store_true')
     parser.set_defaults(output_polymerase_read=False)
+    parser.add_argument("--estimate_np_by_length", help="shall we estimate number of passes by sampled length?", action='store_true')
+    parser.set_defaults(estimate_np_by_length=False)
```
Comment (marghoob) [inline simulate.py]: Do not need this. `action = 'store_true'` implies that it is `False` by default.

Comment (yunfeiguo) [inline simulate.py]: Cool. Thanks!

File: src/main/java/com/bina/lrsim/simulator/Simulator.java [hunk 2]
```diff
@@ -254,21 +274,28 @@ public int simulateWithNoisyAdapter(final String path, final String movieName, f
...
-            if (insertLengths[ii] < sampledReferenceSequence.length) {
-              insertLengths[ii] = sampledReferenceSequence.length;
+        // correct insert lengths if the drawn fragment is shorter, the fractional change might not be realistic, but it avoids crazy coverage in fragment mode
+        // if we want to estimate number of passes by insert lengths, we break up the lengths into chunks
+        // that are equal to lengths of template sequence drawn
+        if (spec.isEstimateNPByLength() && spec.getSequencingMode() == SequencingMode.fragment) {
+          //right now we assume the sampled lengths contain adapter lengths
+          //i.e. each multiples of adapter + template except for the first
+          //one
+          insertLengths = breakUpLengths(insertLengths[0], sampledReferenceSequence.length, Heuristics.SMRT_ADAPTOR_STRING.length);
```
Comment (bayolau) [inline src/main/java/com/bina/lrsim/simulator/Simulator.java]: the first argument should be the sum over `insertLengths`?

File: src/main/java/com/bina/lrsim/simulator/Simulator.java [hunk 3]
```diff
@@ -411,8 +438,7 @@ and assume it starts anywhere completely randomly (if sequencing starts
                 clrBegin = locus.getEnd0() - currentInsertLength;
               else
                 clrEnd = locus.getBegin0() + currentInsertLength;
-            }
-            if (isLastClr) {
+            } else if (isLastClr) {
```
Comment (bayolau) [inline src/main/java/com/bina/lrsim/simulator/Simulator.java]: @yunfeiguo  this is a very very serious bug introduced in refactoring, see https://github.com/bioinform/longislnd/blob/master/src/main/java/com/bina/lrsim/simulator/Simulator.java#L130

we need to stop everything and find out when this happened

File: src/main/java/com/bina/lrsim/simulator/Simulator.java (hunk 2 above)
Comment (yunfeiguo) [inline src/main/java/com/bina/lrsim/simulator/Simulator.java]: The `insertLengths` are summed up here: https://github.com/bioinform/longislnd/pull/31/files#diff-ef9f42154f4ec285ec5b7867845d9ae0R78, right after they are loaded.

Comment (bayolau) [inline src/main/java/com/bina/lrsim/simulator/Simulator.java]: yes, but this forces future dev to read code somewhere else and understand non-local assumptions

Comment (yunfeiguo) [inline src/main/java/com/bina/lrsim/simulator/Simulator.java]: Ok, I will fix this.

File: src/main/java/com/bina/lrsim/simulator/Simulator.java (hunk 3 above)
Comment (yunfeiguo) [inline src/main/java/com/bina/lrsim/simulator/Simulator.java]: Thanks for pointing it out! I have fixed it.

**Comment by yunfeiguo** [comment]
@bayolau `dev` is merged now.

File: src/main/java/com/bina/lrsim/simulator/Simulator.java [hunk 4]
```diff
@@ -127,8 +149,9 @@ private int simulateWithPerfectAdapter(final String path, final String movieName
           final boolean firstClr = insIdx == 0;
           final boolean lastClr = insIdx + 1 == insertLengths.length;
           final int begin = firstClr ? sequence.length - insertLength : 0;
-          final int end = lastClr ? insertLength : sequence.length;
+          final int end = lastClr && (!firstClr) ? insertLength : sequence.length;
```
Comment (bayolau) [inline src/main/java/com/bina/lrsim/simulator/Simulator.java]: good catch thanks

**Comment by bayolau** [comment]
thanks @yunfeiguo

**Comment by yunfeiguo** [comment]
Great. My pleasure.

**Pull request merged**